cohttp-lwt/src/connection.ml

```diff
@@ -195,31 +198,23 @@ struct
       Lwt.return_unit
     | Full (ic, oc)
     | Closing (ic, oc) ->
-      let {req; body; res_r; _} as work = Queue.take connection.waiting in
-      let uri, meth, version, headers =
-        Request.(uri req, meth req, version req, headers req) in
+      let {uri; meth; headers; body; res_r} as work =
+        Queue.take connection.waiting
+      in
 
-      (* select encoding based on header, request, body and make sure
-       * the encoding in the request matches the encoding in the header. *)
+      (* select encoding based on (1st) header or (2nd) body *)
       begin match Header.get_transfer_encoding headers with
         | Unknown ->
-          begin match Request.encoding req with
-          | Fixed 0L
-          (* this is the default selected by Request.make.
-           * XXX: Shouldn't the default be Unknown ? *)
-          | Unknown ->
-            begin match Body.transfer_encoding body with
-            | Fixed _ as e -> Lwt.return (e, body)
-            | Chunked as e when connection.persistent = `True ->
-              Lwt.return (e, body)
-            | Chunked (* connection.persistent <> `True *) ->
-              (* We don't know yet whether chunked encoding is supported.
-               * Therefore use fixed length encoding. *)
-              Body.length body >>= fun (length, body) ->
-              Lwt.return (Cohttp.Transfer.Fixed length, body)
-            | Unknown -> assert false
-            end
-          | e -> Lwt.return (e, body)
+          begin match Body.transfer_encoding body with
+          | Fixed _ as e -> Lwt.return (e, body)
+          | Chunked as e when connection.persistent = `True ->
+            Lwt.return (e, body)
+          | Chunked (* connection.persistent <> `True *) ->
+            (* We don't know yet whether chunked encoding is supported.
+             * Therefore use fixed length encoding. *)
+            Body.length body >>= fun (length, body) ->
+            Lwt.return (Cohttp.Transfer.Fixed length, body)
+          | Unknown -> assert false
```

Comment:
Should we raise an exception in this case?

Reply:
`Body.transfer_encoding` does not return `Unknown`. I can't see how any exception would be more helpful than `Assert_failure`.